### PR TITLE
feat(state): add transparent address indexes to the non-finalized state

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

As part of the address index RPCs, we need to track address balances, UTXOs, and transactions in the non-finalized state.

This PR creates the address indexes in the non-finalized state, but it doesn't connect them to state requests or RPCs yet.

### Specifications

See tickets:
- #3147 
- #3157 
- #3158

### Designs

The `TransparentTransfers` struct holds the address indexes for a single address.

Index updates use `UpdateWith<T> for TransparentTransfers`, where `T` is a tuple of arguments for:
 - created UTXOs, or
 - spent UTXOs

Spent UTXOs are a bit tricky, because we need to add their values to the `ContextuallyValidBlock`.

## Solution

- Create a `TransparentTransfers` address index type
- Implement balance, UTXO, and transaction ID updates for created UTXOs
- Implement balance, UTXO, and transaction ID updates for spent UTXOs
- Remove empty TransparentTransfers after a spend revert
- Add address transfers index accessor methods, which return UTXOs and transactions in chain order
- Make accessor methods provide all the information required for each query

Related changes:
- Add spent UTXO values to `ContextuallyValidBlock`
- Use OrderedUtxos so that spent UTXOs can be stored in chain order
-  Switch to height-based `TransactionLocation`s as part of merging PR #3978

Related bug fixes:
- Stop indexing each transparent output multiple times

Previously, the chain code was indexing each transparent output in the block multiple times (once per transaction). Because we were using `HashMap`s for chain indexes, this bug did not cause any problems. But in the address indexes, it made balances overflow.

(This should also improve non-finalized state performance for large blocks.)

## Review

@conradoplg reviewed a draft of this PR.

We still need to connect these indexes to state requests, but the changes in this PR are enough to get us started.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

## Follow Up Work

In other PRs:
- Implement the address RPCs and state requests
    - Add unit tests
    - Add RPC snapshot tests

### CODE REVIEW

**Comment by teor2345** [comment]
There's something wrong with the balance calculations in this PR:
> The application panicked (crashed).
> Message:  called `Result::unwrap()` on an `Err` value: Constraint { value: 2100025173264129, range: -2100000000000000..=2100000000000000 }
> Location: zebra-state/src/service/non_finalized_state/chain/index.rs:85

**Review by conradoplg — COMMENTED** [review]
I keep reviewing this PR but I can't seem to find the bug :disappointed:  (or any other issue). I'll leave it for someone else to review.

**Comment by teor2345** [comment]
> I keep reviewing this PR but I can't seem to find the bug 😞 (or any other issue). I'll leave it for someone else to review.

I think it's probably a mismatch between the created and spent UTXOs. Probably some missing spent UTXOs, because it's an overflow.

I'll make this part of the code easier to get right.

File: zebra-state/src/service/non_finalized_state/chain.rs [hunk 1]
```diff
@@ -565,18 +600,19 @@ impl UpdateWith<ContextuallyValidBlock> for Chain {
...
-            assert!(
-                prior_pair.is_none(),
+                .insert(transaction_hash, transaction_location);
+            assert_eq!(
+                prior_pair, None,
                 "transactions must be unique within a single chain"
             );
 
-            // add the utxos this produced
-            self.update_chain_tip_with(new_outputs)?;
```
Comment (teor2345) [inline zebra-state/src/service/non_finalized_state/chain.rs:577]: This was an existing bug: `new_outputs` contains all the UTXOs for the block, but it was indexed for every transaction.

File: zebra-state/src/service/non_finalized_state/chain.rs [hunk 2]
```diff
@@ -688,17 +737,19 @@ impl UpdateWith<ContextuallyValidBlock> for Chain {
...
             // remove `transaction.hash` from `tx_by_hash`
             assert!(
                 self.tx_by_hash.remove(transaction_hash).is_some(),
                 "transactions must be present if block was added to chain"
             );
 
-            // remove the utxos this produced
-            self.revert_chain_with(new_outputs, position);
-            // remove the utxos this consumed
-            self.revert_chain_with(inputs, position);
```
Comment (teor2345) [inline zebra-state/src/service/non_finalized_state/chain.rs:700]: This was a bug in this PR: these methods use `tx_by_hash`, but it was reverted before they ran.

**Review by oxarbitrage — DISMISSED** [review]
This is a lot of progress that will unblock a bunch of stuff for at least 3 RPC methods. The pull request is a bit big but a lot of it is reordering. Skipping all that, the new methods this PR brings are pretty clear. I didn't made an in deep review line by line but it looks good to go.

**Comment by teor2345** [comment]
@mergifyio update

**Comment by teor2345** [comment]
@mergifyio update

**Review by oxarbitrage — APPROVED** [review]
Re-approving.

**Pull request merged**